File: dream-server/extensions/services/dashboard-api/routers/updates.py
"""Version checking and update endpoints."""

import asyncio
import json
import logging
import re
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

import httpx
from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException

from config import INSTALL_DIR
from models import VersionInfo, UpdateAction
from security import verify_api_key

logger = logging.getLogger(__name__)

router = APIRouter(tags=["updates"])

_VALID_ACTIONS = {"check", "backup", "update"}

_GITHUB_HEADERS = {"Accept": "application/vnd.github.v3+json"}
_VERSION_CACHE_TTL = 300.0
_version_cache: dict[str, object] = {"expires_at": 0.0, "payload": None}
_version_refresh_task: Optional[asyncio.Task] = None


def _read_current_version() -> str:
    """Read installed version from .env (preferred) or .version file."""
    env_file = Path(INSTALL_DIR) / ".env"
    if env_file.exists():
        try:
            for line in env_file.read_text().splitlines():
                if line.startswith("DREAM_VERSION="):
                    return line.split("=", 1)[1].strip().strip("\"'")
        except OSError:
            pass
    version_file = Path(INSTALL_DIR) / ".version"
    if version_file.exists():
        try:
            raw = version_file.read_text().strip()
            if raw:
                return raw
        except OSError:
            pass
    manifest_file = Path(INSTALL_DIR) / "manifest.json"
    if manifest_file.exists():
        try:
            data = json.loads(manifest_file.read_text())
            version = (
                data.get("release", {}).get("version")
                or data.get("dream_version")
                or data.get("manifestVersion")
            )
            if version:
                return str(version)
        except (OSError, json.JSONDecodeError, ValueError, AttributeError):
            pass
    main_file = Path(__file__).resolve().parents[1] / "main.py"
    if main_file.exists():
        try:
            match = re.search(r'version\s*=\s*"([^"]+)"', main_file.read_text())
            if match:
                return match.group(1)
        except OSError:
            pass
    return "0.0.0"


def _get_cached_release_payload(allow_stale: bool = False) -> Optional[dict]:
    payload = _version_cache.get("payload")
    if payload is None:
        return None
    if allow_stale or time.monotonic() < float(_version_cache.get("expires_at", 0.0)):
        return payload  # type: ignore[return-value]
    return None


def _build_version_result(current: str, payload: Optional[dict]) -> dict:
    result = {
        "current": current,
        "latest": None,
        "update_available": False,
        "changelog_url": None,
        "checked_at": datetime.now(timezone.utc).isoformat() + "Z",
    }
    if not payload:
        return result

    latest = (payload.get("latest") or "").lstrip("v")
    if not latest:
        return result

    result["latest"] = latest
    result["changelog_url"] = payload.get("changelog_url")
    result["checked_at"] = payload.get("checked_at") or result["checked_at"]

    current_parts = [int(x) for x in current.split(".") if x.isdigit()][:3]
    latest_parts = [int(x) for x in latest.split(".") if x.isdigit()][:3]
    current_parts += [0] * (3 - len(current_parts))
    latest_parts += [0] * (3 - len(latest_parts))
    result["update_available"] = latest_parts > current_parts
    return result


async def _refresh_release_cache() -> Optional[dict]:
    global _version_cache
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            response = await client.get(
                "https://api.github.com/repos/Light-Heart-Labs/DreamServer/releases/latest",
                headers=_GITHUB_HEADERS,
            )
        data = response.json()
        payload = {
            "latest": data.get("tag_name", "").lstrip("v"),
            "changelog_url": data.get("html_url"),
            "checked_at": datetime.now(timezone.utc).isoformat() + "Z",
        }
        _version_cache = {
            "expires_at": time.monotonic() + _VERSION_CACHE_TTL,
            "payload": payload,
        }
        return payload
    except (httpx.HTTPError, httpx.TimeoutException, json.JSONDecodeError, OSError, ValueError):
        return _get_cached_release_payload(allow_stale=True)


def _ensure_release_refresh() -> asyncio.Task:
    global _version_refresh_task
    if _version_refresh_task is None or _version_refresh_task.done():
        _version_refresh_task = asyncio.create_task(_refresh_release_cache())
    return _version_refresh_task


@router.get("/api/version", response_model=VersionInfo, dependencies=[Depends(verify_api_key)])
async def get_version():
    """Get current Dream Server version without blocking page load on GitHub."""
    current = await asyncio.to_thread(_read_current_version)
    cached = _get_cached_release_payload()
    if cached:
        return _build_version_result(current, cached)

    stale = _get_cached_release_payload(allow_stale=True)
    refresh_task = _ensure_release_refresh()

    if stale:
        return _build_version_result(current, stale)

    try:
        payload = await asyncio.wait_for(asyncio.shield(refresh_task), timeout=1.25)
        return _build_version_result(current, payload)
    except asyncio.TimeoutError:
        logger.debug("Version refresh still in progress; returning local version immediately")
        return _build_version_result(current, None)


@router.get("/api/releases/manifest", dependencies=[Depends(verify_api_key)])
async def get_release_manifest():
    """Get release manifest with version history (non-blocking)."""
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            resp = await client.get(
                "https://api.github.com/repos/Light-Heart-Labs/DreamServer/releases?per_page=5",
                headers=_GITHUB_HEADERS,
            )
        releases = resp.json()
        if not isinstance(releases, list):
            raise httpx.HTTPError(f"unexpected releases response: {type(releases).__name__}")
        return {
            "releases": [
                {"version": r.get("tag_name", "").lstrip("v"), "date": r.get("published_at", ""), "title": r.get("name", ""), "changelog": r.get("body", "")[:500] + "..." if len(r.get("body", "")) > 500 else r.get("body", ""), "url": r.get("html_url", ""), "prerelease": r.get("prerelease", False)}
                for r in releases
            ],
            "checked_at": datetime.now(timezone.utc).isoformat() + "Z"
        }
    except (httpx.HTTPError, httpx.TimeoutException, json.JSONDecodeError, OSError):
        current = await asyncio.to_thread(_read_current_version)
        return {
            "releases": [{"version": current, "date": datetime.now(timezone.utc).isoformat() + "Z", "title": f"Dream Server {current}", "changelog": "Release information unavailable. Check GitHub directly.", "url": "https://github.com/Light-Heart-Labs/DreamServer/releases", "prerelease": False}],
            "checked_at": datetime.now(timezone.utc).isoformat() + "Z",
            "error": "Could not fetch release information"
        }


_UPDATE_ENV_KEYS = {
    "DREAM_VERSION", "TIER", "LLM_MODEL", "GGUF_FILE",
    "CTX_SIZE", "GPU_BACKEND", "N_GPU_LAYERS",
}


@router.get("/api/update/dry-run", dependencies=[Depends(verify_api_key)])
async def get_update_dry_run():
    """Preview what a dream update would change without applying anything.

    Returns version comparison, configured image tags, and the .env keys
    that the update process reads or writes.  No containers are started,
    stopped, or re-created.
    """
    import urllib.request
    import urllib.error

    install_path = Path(INSTALL_DIR)

    # ── current version ───────────────────────────────────────────────────────
    current = "0.0.0"
    env_file = install_path / ".env"
    version_file = install_path / ".version"

    if env_file.exists():
        for line in env_file.read_text().splitlines():
            if line.startswith("DREAM_VERSION="):
                current = line.split("=", 1)[1].strip()
                break
    if current == "0.0.0" and version_file.exists():
        try:
            raw = version_file.read_text().strip()
            parsed = json.loads(raw) if raw.startswith("{") else None
            current = (parsed or {}).get("version", raw) or raw or "0.0.0"
        except (json.JSONDecodeError, OSError):
            pass

    # ── latest version from GitHub ────────────────────────────────────────────
    latest: Optional[str] = None
    changelog_url: Optional[str] = None
    update_available = False

    try:
        req = urllib.request.Request(
            "https://api.github.com/repos/Light-Heart-Labs/DreamServer/releases/latest",
            headers={"Accept": "application/vnd.github.v3+json"},
        )
        with urllib.request.urlopen(req, timeout=8) as resp:
            data = json.loads(resp.read())
            latest = data.get("tag_name", "").lstrip("v") or None
            changelog_url = data.get("html_url") or None
            if latest:
                def _parts(v: str) -> list[int]:
                    return ([int(x) for x in v.split(".") if x.isdigit()][:3] + [0, 0, 0])[:3]
                update_available = _parts(latest) > _parts(current)
    except (urllib.error.URLError, urllib.error.HTTPError, OSError, json.JSONDecodeError, ValueError):
        pass

    # ── configured image tags from compose files ──────────────────────────────
    images: list[str] = []
    for compose_file in sorted(install_path.glob("docker-compose*.yml")):
        try:
            for line in compose_file.read_text().splitlines():
                stripped = line.strip()
                if stripped.startswith("image:"):
                    tag = stripped.split(":", 1)[1].strip()
                    if tag and tag not in images:
                        images.append(tag)
        except OSError:
            pass

    # ── .env keys relevant to the update path ────────────────────────────────
    env_snapshot: dict[str, str] = {}
    if env_file.exists():
        for line in env_file.read_text().splitlines():
            line = line.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            key, _, val = line.partition("=")
            if key in _UPDATE_ENV_KEYS:
                env_snapshot[key] = val

    return {
        "dry_run": True,
        "current_version": current,
        "latest_version": latest,
        "update_available": update_available,
        "changelog_url": changelog_url,
        "images": images,
        "env_keys": env_snapshot,
    }


@router.post("/api/update")
async def trigger_update(action: UpdateAction, background_tasks: BackgroundTasks, api_key: str = Depends(verify_api_key)):
    """Trigger update actions via dashboard."""
    if action.action not in _VALID_ACTIONS:
        raise HTTPException(status_code=400, detail=f"Unknown action: {action.action}")

    script_path = Path(INSTALL_DIR).parent / "scripts" / "dream-update.sh"
    if not script_path.exists():
        install_script = Path(INSTALL_DIR) / "install.sh"
        if install_script.exists():
            script_path = Path(INSTALL_DIR).parent / "scripts" / "dream-update.sh"
        else:
            script_path = Path(INSTALL_DIR) / "scripts" / "dream-update.sh"

    if not script_path.exists():
        logger.error("dream-update.sh not found at %s", script_path)
        raise HTTPException(status_code=501, detail="Update system not installed.")

    if action.action == "check":
        try:
            proc = await asyncio.create_subprocess_exec(
                str(script_path), "check",
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=30)
            return {"success": True, "update_available": proc.returncode == 2, "output": stdout.decode() + stderr.decode()}
        except asyncio.TimeoutError:
            raise HTTPException(status_code=504, detail="Update check timed out")
        except OSError:
            logger.exception("Update check failed")
            raise HTTPException(status_code=500, detail="Check failed")
    elif action.action == "backup":
        try:
            proc = await asyncio.create_subprocess_exec(
                str(script_path), "backup", f"dashboard-{datetime.now().strftime('%Y%m%d-%H%M%S')}",
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=60)
            return {"success": proc.returncode == 0, "output": stdout.decode() + stderr.decode()}
        except asyncio.TimeoutError:
            raise HTTPException(status_code=504, detail="Backup timed out")
        except OSError:
            logger.exception("Backup failed")
            raise HTTPException(status_code=500, detail="Backup failed")
    elif action.action == "update":
        async def run_update():
            proc = await asyncio.create_subprocess_exec(
                str(script_path), "update",
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            await proc.communicate()
        background_tasks.add_task(run_update)
        return {"success": True, "message": "Update started in background. Check logs for progress."}

File: dream-server/extensions/services/dashboard-api/crates/dashboard-api/src/routes/updates.rs
//! Updates router — /api/updates/* endpoints. Mirrors routers/updates.py.

use axum::extract::State;
use axum::Json;
use serde_json::{json, Value};
use std::path::PathBuf;

use crate::state::AppState;

fn install_dir() -> String {
    std::env::var("DREAM_INSTALL_DIR")
        .unwrap_or_else(|_| shellexpand::tilde("~/dream-server").to_string())
}

/// GET /api/updates/version — current version info
pub async fn version_info() -> Json<Value> {
    let install = install_dir();
    let version_file = PathBuf::from(&install).join("VERSION");
    let current = std::fs::read_to_string(&version_file)
        .ok()
        .map(|s| s.trim().to_string())
        .unwrap_or_else(|| "unknown".to_string());

    Json(json!({
        "current": current,
        "latest": null,
        "update_available": false,
        "changelog_url": null,
        "checked_at": null,
    }))
}

/// POST /api/updates/check — check for available updates
pub async fn check_updates(State(state): State<AppState>) -> Json<Value> {
    let install = install_dir();
    let version_file = PathBuf::from(&install).join("VERSION");
    let current = std::fs::read_to_string(&version_file)
        .ok()
        .map(|s| s.trim().to_string())
        .unwrap_or_else(|| "unknown".to_string());

    // Try to fetch latest version from GitHub releases
    let latest = match state
        .http
        .get("https://api.github.com/repos/Light-Heart-Labs/DreamServer/releases/latest")
        .header("User-Agent", "DreamServer-Dashboard")
        .send()
        .await
    {
        Ok(resp) if resp.status().is_success() => {
            let data: Value = resp.json().await.unwrap_or(json!({}));
            data["tag_name"].as_str().map(|s| s.trim_start_matches('v').to_string())
        }
        _ => None,
    };

    let update_available = latest
        .as_ref()
        .map(|l| l != &current)
        .unwrap_or(false);

    Json(json!({
        "current": current,
        "latest": latest,
        "update_available": update_available,
        "changelog_url": latest.as_ref().map(|_| "https://github.com/Light-Heart-Labs/DreamServer/releases"),
        "checked_at": chrono::Utc::now().to_rfc3339(),
    }))
}

/// GET /api/releases/manifest — release manifest with version history
pub async fn releases_manifest(State(state): State<AppState>) -> Json<Value> {
    match state
        .http
        .get("https://api.github.com/repos/Light-Heart-Labs/DreamServer/releases?per_page=5")
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "DreamServer-Dashboard")
        .send()
        .await
    {
        Ok(resp) if resp.status().is_success() => {
            let releases: Vec<Value> = resp.json().await.unwrap_or_default();
            let formatted: Vec<Value> = releases
                .iter()
                .map(|r| {
                    let body = r["body"].as_str().unwrap_or("");
                    let truncated = if body.len() > 500 {
                        format!("{}...", &body[..500])
                    } else {
                        body.to_string()
                    };
                    json!({
                        "version": r["tag_name"].as_str().unwrap_or("").trim_start_matches('v'),
                        "date": r["published_at"],
                        "title": r["name"],
                        "changelog": truncated,
                        "url": r["html_url"],
                        "prerelease": r["prerelease"],
                    })
                })
                .collect();
            Json(json!({
                "releases": formatted,
                "checked_at": chrono::Utc::now().to_rfc3339(),
            }))
        }
        _ => {
            let current = std::fs::read_to_string(
                PathBuf::from(&install_dir()).join("VERSION"),
            )
            .ok()
            .map(|s| s.trim().to_string())
            .unwrap_or_else(|| "unknown".to_string());
            Json(json!({
                "releases": [{
                    "version": current,
                    "date": chrono::Utc::now().to_rfc3339(),
                    "title": format!("Dream Server {current}"),
                    "changelog": "Release information unavailable. Check GitHub directly.",
                    "url": "https://github.com/Light-Heart-Labs/DreamServer/releases",
                    "prerelease": false,
                }],
                "checked_at": chrono::Utc::now().to_rfc3339(),
                "error": "Could not fetch release information",
            }))
        }
    }
}

/// GET /api/update/dry-run — preview what an update would change
pub async fn update_dry_run() -> Json<Value> {
    let install = install_dir();
    let install_path = PathBuf::from(&install);

    // Read current version from .env or .version
    let mut current = "0.0.0".to_string();
    let env_file = install_path.join(".env");
    if env_file.exists() {
        if let Ok(text) = std::fs::read_to_string(&env_file) {
            for line in text.lines() {
                if let Some(val) = line.strip_prefix("DREAM_VERSION=") {
                    current = val.trim().trim_matches('"').trim_matches('\'').to_string();
                    break;
                }
            }
        }
    }
    if current == "0.0.0" {
        let version_file = install_path.join(".version");
        if let Ok(text) = std::fs::read_to_string(&version_file) {
            let trimmed = text.trim();
            if !trimmed.is_empty() {
                current = trimmed.to_string();
            }
        }
    }

    // Configured image tags from compose files
    let mut images: Vec<String> = Vec::new();
    if let Ok(entries) = std::fs::read_dir(&install_path) {
        let mut compose_files: Vec<_> = entries
            .filter_map(|e| e.ok())
            .filter(|e| {
                let name = e.file_name().to_string_lossy().to_string();
                name.starts_with("docker-compose") && name.ends_with(".yml")
            })
            .collect();
        compose_files.sort_by_key(|e| e.file_name());
        for entry in compose_files {
            if let Ok(text) = std::fs::read_to_string(entry.path()) {
                for line in text.lines() {
                    let stripped = line.trim();
                    if let Some(tag) = stripped.strip_prefix("image:") {
                        let tag = tag.trim().to_string();
                        if !tag.is_empty() && !images.contains(&tag) {
                            images.push(tag);
                        }
                    }
                }
            }
        }
    }

    // .env keys relevant to update
    let update_keys: std::collections::HashSet<&str> = [
        "DREAM_VERSION", "TIER", "LLM_MODEL", "GGUF_FILE",
        "CTX_SIZE", "GPU_BACKEND", "N_GPU_LAYERS",
    ]
    .into_iter()
    .collect();

    let mut env_snapshot = serde_json::Map::new();
    if let Ok(text) = std::fs::read_to_string(&env_file) {
        for line in text.lines() {
            let line = line.trim();
            if line.is_empty() || line.starts_with('#') || !line.contains('=') {
                continue;
            }
            if let Some((key, val)) = line.split_once('=') {
                if update_keys.contains(key) {
                    env_snapshot.insert(key.to_string(), json!(val));
                }
            }
        }
    }

    Json(json!({
        "dry_run": true,
        "current_version": current,
        "latest_version": null,
        "update_available": false,
        "changelog_url": null,
        "images": images,
        "env_keys": env_snapshot,
    }))
}

/// POST /api/update — trigger update action (check, backup, update)
pub async fn update_action(Json(body): Json<Value>) -> Json<Value> {
    let action = body["action"].as_str().unwrap_or("check");
    match action {
        "check" => Json(json!({"status": "ok", "action": "check", "message": "Use POST /api/updates/check"})),
        "backup" => {
            // Trigger backup via dream-cli
            match tokio::process::Command::new("dream-cli")
                .args(["backup", "create"])
                .output()
                .await
            {
                Ok(output) if output.status.success() => {
                    Json(json!({"status": "ok", "action": "backup", "message": "Backup created"}))
                }
                Ok(output) => {
                    let stderr = String::from_utf8_lossy(&output.stderr);
                    Json(json!({"status": "error", "action": "backup", "message": format!("Backup failed: {stderr}")}))
                }
                Err(e) => Json(json!({"status": "error", "action": "backup", "message": format!("Failed to run dream-cli: {e}")})),
            }
        }
        "update" => {
            Json(json!({"status": "error", "action": "update", "message": "In-place updates not yet supported via API"}))
        }
        _ => Json(json!({"status": "error", "message": format!("Unknown action: {action}")})),
    }
}

#[cfg(test)]
mod tests {
    use crate::state::AppState;
    use axum::body::Body;
    use http::Request;
    use http_body_util::BodyExt;
    use serde_json::Value;
    use std::collections::HashMap;
    use tower::ServiceExt;

    fn app() -> axum::Router {
        crate::build_router(AppState::new(
            HashMap::new(), vec![], vec![], "test-key".into(),
        ))
    }

    fn auth_header() -> String {
        "Bearer test-key".to_string()
    }

    #[tokio::test]
    async fn version_info_requires_auth() {
        let req = Request::builder()
            .uri("/api/version")
            .body(Body::empty())
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 401);
    }

    #[tokio::test]
    async fn version_info_returns_shape() {
        let req = Request::builder()
            .uri("/api/version")
            .header("authorization", auth_header())
            .body(Body::empty())
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert!(val.get("current").is_some());
        assert!(val.get("update_available").is_some());
    }

    #[tokio::test]
    async fn update_dry_run_returns_shape() {
        let req = Request::builder()
            .uri("/api/update/dry-run")
            .header("authorization", auth_header())
            .body(Body::empty())
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(val["dry_run"], true);
        assert!(val.get("current_version").is_some());
        assert!(val.get("images").is_some());
    }

    #[tokio::test]
    async fn update_action_unknown_returns_error() {
        let req = Request::builder()
            .method("POST")
            .uri("/api/update")
            .header("authorization", auth_header())
            .header("content-type", "application/json")
            .body(Body::from(r#"{"action":"invalid"}"#))
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(val["status"], "error");
    }

    #[tokio::test]
    async fn update_action_requires_auth() {
        let req = Request::builder()
            .method("POST")
            .uri("/api/update")
            .header("content-type", "application/json")
            .body(Body::from(r#"{"action":"check"}"#))
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 401);
    }

    #[tokio::test]
    async fn releases_manifest_requires_auth() {
        let req = Request::builder()
            .uri("/api/releases/manifest")
            .body(Body::empty())
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 401);
    }

    #[tokio::test]
    async fn releases_manifest_returns_releases_key() {
        let req = Request::builder()
            .uri("/api/releases/manifest")
            .header("authorization", auth_header())
            .body(Body::empty())
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert!(val.get("releases").is_some());
        assert!(val["releases"].is_array());
        assert!(val.get("checked_at").is_some());
    }

    #[tokio::test]
    async fn update_action_backup_returns_error_without_cli() {
        let req = Request::builder()
            .method("POST")
            .uri("/api/update")
            .header("authorization", auth_header())
            .header("content-type", "application/json")
            .body(Body::from(r#"{"action":"backup"}"#))
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(val["action"], "backup");
        // dream-cli not available in test env => error
        assert_eq!(val["status"], "error");
    }

    #[tokio::test]
    async fn update_action_update_returns_not_supported() {
        let req = Request::builder()
            .method("POST")
            .uri("/api/update")
            .header("authorization", auth_header())
            .header("content-type", "application/json")
            .body(Body::from(r#"{"action":"update"}"#))
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(val["status"], "error");
        assert_eq!(val["action"], "update");
    }

    #[tokio::test]
    async fn dry_run_requires_auth() {
        let req = Request::builder()
            .uri("/api/update/dry-run")
            .body(Body::empty())
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 401);
    }

    #[tokio::test]
    async fn update_action_check_returns_redirect() {
        let req = Request::builder()
            .method("POST")
            .uri("/api/update")
            .header("authorization", auth_header())
            .header("content-type", "application/json")
            .body(Body::from(r#"{"action":"check"}"#))
            .unwrap();
        let resp = app().oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(val["action"], "check");
        assert_eq!(val["status"], "ok");
    }
}
